tests/test_request_handling.py

```diff
@@ -0,0 +1,525 @@
+            param="tool_choice",
+        )
+    if request.functions is None and request.function_call is not None:
+        return create_error_response(
+            status_code=HTTPStatus.BAD_REQUEST,
+            message=f"Invalid value for 'function_call': 'function_call' is only allowed when 'functions' are specified.",
+            param="function_call",
+        )
+    if request.tools is None and request.tool_choice is not None:
+        return create_error_response(
+            status_code=HTTPStatus.BAD_REQUEST,
+            message=f"Invalid value for 'tool_choice': 'tool_choice' is only allowed when 'tools' are specified.",
+            param="tool_choice",
+        )
+    return
+
+
+def convert_jsonresponse_to_json(obj):
+    return json.loads(obj.body.decode())
+
+
+def get_token_ids_from_request(
+    request: ChatCompletionRequest, tokenizer: AutoTokenizer
+):
+    if request.tools:
+        tools = enforce_tool_choice(
+            choice=request.tool_choice, tools_or_functions=request.tools
+        )
+        tools_or_functions = [item.dict() for item in tools]
+        tool_func_choice = request.tool_choice if request.tool_choice else "auto"
+    elif request.functions:
+        functions = enforce_tool_choice(
+            choice=request.function_call, tools_or_functions=request.functions
+        )
+        tools_or_functions = [item.dict() for item in functions]
+        tool_func_choice = request.function_call if request.function_call else "auto"
+    else:
+        tools_or_functions = []
+        tool_func_choice = "none"
+
+    prompt_token_ids = prepare_messages_for_inference(
+        tokenizer=tokenizer,
+        messages=request.messages,
+        tools_or_functions=tools_or_functions,
+        tool_choice=tool_func_choice,
+    ).tolist()[0]
+
+    return prompt_token_ids
+
+
+class TestRequestHandling(unittest.IsolatedAsyncioTestCase):
+    def __init__(self, *args, **kwargs):
+        super(TestRequestHandling, self).__init__(*args, **kwargs)
+        self.served_model = "meetkai/functionary-small-v2.5"
+        self.default_messages = [{"role": "user", "content": "How are you?"}]
+        self.default_functions = [
+            {
+                "name": "get_current_weather",
+                "description": "Get the current weather",
+                "parameters": {
+                    "type": "object",
+                    "properties": {
+                        "location": {
+                            "type": "string",
+                            "description": "The city and state, e.g. San Francisco, CA",
+                        }
+                    },
+                    "required": ["location"],
+                },
+            },
+        ]
+        self.default_tools = [
+            {"type": "function", "function": self.default_functions[0]}
+        ]
+        self.tokenizers_to_test = [
```

Comment:
Actually I think tokenizers here are used to create prompt_template, right ? so why don't we just use 

self.test_template_versions = get_available_templates_versions() --> list of string in prompte_template/__init__.py 
so we don't need to care about downloading tokenizer from HF and also private model that we don't set public?